### PR TITLE
fix(bug): limit fix

### LINKED ISSUE #1151: bug(python): Passing limit of `None` or 0 or -1 to `search().limit()` fails
### LanceDB version

v0.6.5

### What happened?

Passing limit of `None` or 0 or -1 to `search().limit()` fails.
It should have returned all results (no limit).

### Are there known steps to reproduce?

```python
import lancedb
import numpy as np
import pyarrow as pa


embed_len = 3
lance_db = lancedb.connect("db/test-lance-0")
table = lance_db.create_table(
    "documents",
    schema=pa.schema(
        [pa.field("text", pa.utf8()), pa.field("embed", pa.list_(pa.float16(), embed_len))]
    ),
)
table.add([dict(text="Prime minister", embed=np.ones([embed_len], dtype=np.float16))])
table.create_fts_index(["text"])
table.checkout_latest()

# These work
print(table.search(np.ones([embed_len], dtype=np.float16), "embed").limit(1).to_list())
print(table.search("minister", "text").limit(1).to_list())

# These all fail
print(table.search(np.ones([embed_len], dtype=np.float16), "embed").limit(None).to_list())
print(table.search(np.ones([embed_len], dtype=np.float16), "embed").limit(0).to_list())
print(table.search(np.ones([embed_len], dtype=np.float16), "embed").limit(-1).to_list())
# Traceback (most recent call last):
#   File "test_lance.py", line 28, in <module>
#     print(table.search(np.ones([embed_len], dtype=np.float16), "embed").limit(0).to_list())
#   File ".../lib/python3.10/site-packages/lancedb/query.py", line 302, in to_list
#     return self.to_arrow().to_pylist()
#   File ".../lib/python3.10/site-packages/lancedb/query.py", line 527, in to_arrow
#     return self.to_batches().read_all()
#   File ".../lib/python3.10/site-packages/lancedb/query.py", line 545, in to_batches
#     query = Query(
#   File ".../lib/python3.10/site-packages/pydantic/main.py", line 164, in __init__
#     __pydantic_self__.__pydantic_validator__.validate_python(data, self_instance=__pydantic_self__)
# pydantic_core._pydantic_core.ValidationError: 1 validation error for Query
# k
#   Input should be a valid integer [type=int_type, input_value=None, input_type=NoneType]
#     For further information visit https://errors.pydantic.dev/2.4/v/int_type
print(table.search("minister", "text").limit(None).to_list())
print(table.search("minister", "text").limit(0).to_list())
print(table.search("minister", "text").limit(-1).to_list())
# Traceback (most recent call last):
#   File "test_lance.py", line 45, in <module>
#     print(table.search("minister", "text").limit(None).to_list())
#   File ".../lib/python3.10/site-packages/lancedb/query.py", line 302, in to_list
#     return self.to_arrow().to_pylist()
#   File ".../lib/python3.10/site-packages/lancedb/query.py", line 682, in to_arrow
#     row_ids, scores = search_index(
#   File ".../lib/python3.10/site-packages/lancedb/fts.py", line 195, in search_index
#     results = searcher.search(query, limit)
# TypeError: argument 'limit': 'NoneType' object cannot be interpreted as an integer

```

### PR DESCRIPTION
**Pull request by gagan-bhullar-tech**
PR fixes #1151

### CODE REVIEW

**Comment by gagan-bhullar-tech** [comment]
@wjones127 can you please review

File: python/python/tests/test_table.py [hunk 1]
```diff
@@ -756,9 +756,9 @@ def test_empty_query(db):
     df = table.search().select(["id"]).to_pandas()
     assert len(df) == 10
     df = table.search().select(["id"]).limit(None).to_pandas()
-    assert len(df) == 100
+    assert len(df) == 10
     df = table.search().select(["id"]).limit(-1).to_pandas()
-    assert len(df) == 100
+    assert len(df) == 10
```
Comment (wjones127) [inline python/python/tests/test_table.py:761]: This is a regression. Passing `None` should mean return all rows, not default to 10. This test should be left as is.

**Review by wjones127 — COMMENTED** [review]
Hi @gagan-bhullar-tech . Thanks for making a PR.

I think you misunderstand the user ask in the original issue. They were complaining that you couldn't remove the limit and get all rows back for ANN queries. But this PR does the opposite: you made is so that if you remove the limit it defaults to the limit of 10 for *all queries*. That's definitely not what we want.

The Lance API requires the `k` parameter, so if we wanted to actually remove this limit we could compute `table.count_rows()` and pass that as the limit. (Because of the way IVF_PQ queries work, this still won't return all results in order unless you set `n_probes` equal to all rows.)

TBH, I'm not sure if this issue is worth fixing anyways. It's easy to work around (pass a really high limit) and having no limit doesn't have much of an application.

**Comment by wjones127** [comment]
Thinking about it, the solution I'd like to see if that when it is a ANN / KNN query, we should have a clear error message saying that `limit` is required for those queries. Discussed in detail here: https://github.com/lancedb/lancedb/issues/1151#issuecomment-2302932158

**Comment by gagan-bhullar-tech** [comment]
@wjones127 thanks. I was checking the code you mentioned in the comment. `self` does not contain `_nearest` property. Can you please let me know how to find out whether it is ANN or KNN query.

**Comment by gagan-bhullar-tech** [comment]
@wjones127 thanks. Can you please review.

**Comment by gagan-bhullar-tech** [comment]
@wjones127 thanks. Can you please review.

**Review by wjones127 — APPROVED** [review]
This looks good. Thanks!

**Pull request merged**